File: scripts/scenarios/scenarios/52_rate_limit_behavior.py
"""Scenario 52: "The Rate Limit Dance" — Rate Limit Client Behavior

Luna hits rate limits on multiple endpoints, verifies 429 responses
with Retry-After headers, and confirms recovery after cooldown.

Services: PDS
"""

# SPDX-FileCopyrightText: 2025-2026 Jack Valinsky
# SPDX-License-Identifier: Unlicense OR CC0-1.0

from __future__ import annotations

import sys
import time
from pathlib import Path

_project_root = str(Path(__file__).resolve().parent.parent.parent.parent)
if _project_root not in sys.path:
    sys.path.insert(0, _project_root)

from scripts.lib.atproto import (
    XrpcClient, get_character, PDS1, ScenarioResult, timed_call,
    create_account_or_login, now_iso,
)


def run() -> ScenarioResult:
    result = ScenarioResult("Rate Limit Client Behavior")
    result.start()

    pds = XrpcClient(PDS1)
    luna = get_character("luna")

    timed_call(result, "PDS health check",
               lambda: pds.wait_for_healthy(timeout=30))
    if result.failed > 0:
        result.finish()
        return result

    # Create account
    session = timed_call(
        result, "Create account for Luna",
        lambda: create_account_or_login(pds, luna),
        detail_fn=lambda s: f"did={s.get('did', '?')}",
    )

    if not session:
        result.finish()
        return result

    luna.did = session["did"]
    luna.access_jwt = session["accessJwt"]

    # Rapidly create records to hit rate limits
    rate_limited = False
    retry_after = None
    responses = []

    timed_call(
        result, "Rapidly create records to trigger rate limit",
        lambda: _rapid_create_records(pds, luna, responses),
    )

    # Analyze responses for rate limiting
    for resp in responses:
        status = resp.get("status", 0)
        if status == 429:
            rate_limited = True
            retry_after = resp.get("headers", {}).get("retry-after")
            result.step_passed("Rate limit triggered",
                               f"status=429, retry_after={retry_after}")
            break

    if not rate_limited:
        result.step_skipped("Rate limit trigger",
                            "No 429 response received (limits may be high for local dev)")

    # Verify Retry-After header
    if rate_limited and retry_after:
        try:
            retry_seconds = int(retry_after)
            result.step_passed("Retry-After header valid",
                               f"retry_after={retry_seconds}s")
        except (ValueError, TypeError):
            result.step_skipped("Retry-After header", f"non-numeric: {retry_after}")

    # Test rate limits on uploadBlob
    upload_responses = []
    timed_call(
        result, "Rapidly upload blobs to trigger rate limit",
        lambda: _rapid_upload_blobs(pds, luna, upload_responses),
    )

    upload_rate_limited = any(r.get("status") == 429 for r in upload_responses)
    if upload_rate_limited:
        result.step_passed("Upload rate limit triggered", "status=429")
    else:
        result.step_skipped("Upload rate limit", "No 429 on upload (limits may be high)")

    # Wait briefly and verify recovery
    time.sleep(2)

    # Verify requests succeed after cooldown
    recovery_post = {
        "$type": "app.bsky.feed.post",
        "text": "Post after rate limit cooldown",
        "createdAt": now_iso(),
    }

    timed_call(
        result, "Verify requests succeed after cooldown",
        lambda: pds.repositories.create_record(
            luna.did, "app.bsky.feed.post",
            f"recovery-{int(time.time())}", recovery_post, luna.access_jwt,
        ),
        detail_fn=lambda r: f"recovered={r is not None}",
    )

    # Verify session is still valid
    timed_call(
        result, "Verify session still valid",
        lambda: pds.accounts.get_session(luna.access_jwt),
    )

    result.finish()
    return result


def _rapid_create_records(pds, luna, responses, count=20):
    """Rapidly create records to trigger rate limiting."""
    for i in range(count):
        try:
            rkey = f"rate-test-{i}-{int(time.time())}"
            record = {
                "$type": "app.bsky.feed.post",
                "text": f"Rate limit test post {i}",
                "createdAt": now_iso(),
            }
            resp = pds.repositories.create_record(
                luna.did, "app.bsky.feed.post", rkey,
                record, luna.access_jwt,
            )
            responses.append({"status": 200, "data": resp})
        except Exception as e:
            error_str = str(e)
            status = 429 if "429" in error_str or "rate" in error_str.lower() else 0
            responses.append({"status": status, "error": error_str})
            if status == 429:
                break  # Stop once rate limited


def _rapid_upload_blobs(pds, luna, responses, count=10):
    """Rapidly upload blobs to trigger rate limiting."""
    import base64
    png_data = base64.b64decode(
        "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg=="
    )

    for i in range(count):
        try:
            resp = pds.repositories.upload_blob(
                luna.did, png_data, "image/png", luna.access_jwt,
            )
            responses.append({"status": 200, "data": resp})
        except Exception as e:
            error_str = str(e)
            status = 429 if "429" in error_str or "rate" in error_str.lower() else 0
            responses.append({"status": status, "error": error_str})
            if status == 429:
                break

File: scripts/scenarios/scenarios/52_rate_limit_behavior.ts
/**
 * @module scenarios/52_rate_limit_behavior
 *
 * Scenario: Triggers rate limiting on record creation and verifies recovery.
 *
 * Behavior:
 * - Executes the 52 rate limit behavior scenario.
 * - Validates core operations.
 *
 * Expectations:
 * - Scenario completes successfully without errors.
 */

import { getActor, PDS1 } from "../../lib/deno/config.ts";
import { ScenarioResult } from "../../lib/deno/runner.ts";
export { ScenarioResult, StepResult, StepStatus } from "../../lib/deno/runner.ts";
export type { ScenarioReport } from "../../lib/deno/runner.ts";
import { XrpcClient, XrpcError } from "../../lib/deno/client.ts";
import { assert } from "../../lib/deno/assertions.ts";
import { timedCall } from "../../lib/deno/runner.ts";

/**
 * Executes the scenario logic.
 * @returns A promise that resolves to the scenario result
 */

function now() {
  return new Date().toISOString();
}

function rateLimitRecoveryDelayMs(): number {
  const configuredWindow = Number(Deno.env.get("PDS_RATELIMIT_DID_WINDOW") ?? "0");
  const seconds = Number.isFinite(configuredWindow) && configuredWindow > 0
    ? configuredWindow + 1
    : 65;
  return Math.min(Math.max(seconds, 3), 90) * 1000;
}

export async function run(): Promise<ScenarioResult> {
  const result = new ScenarioResult("Rate Limit Client Behavior");
  result.start();

  const pds = new XrpcClient(PDS1);
  const luna = getActor("luna");

  await timedCall(result, "PDS health check", async () => {
    await pds.waitForHealthy(30);
  });

  if (result.failed > 0) return result;

  const session = await pds.accounts.createAccount(luna.handle, luna.email, luna.password).catch(
    () => pds.accounts.createSession(luna.handle, luna.password),
  );

  if (!session) {
    result.stepFailed("Setup", "Failed to obtain session");
    result.finish();
    return result;
  }
  luna.did = session.did;
  luna.accessJwt = session.accessJwt;

  // Rapidly create records
  let hitLimit = false;
  for (let i = 0; i < 100; i++) {
    try {
      await pds.records.createRecord(luna.did, "app.bsky.feed.post", {
        $type: "app.bsky.feed.post",
        text: `Rate test ${i}`,
        createdAt: now(),
      }, luna.accessJwt);
    } catch (e) {
      if (e instanceof XrpcError && e.status === 429) {
        hitLimit = true;
        result.stepPassed("Rate limit triggered (429)");
        break;
      }
    }
  }

  if (!hitLimit) result.stepSkipped("Rate limit trigger", "Limits might be too high for local dev");

  // Recovery
  await new Promise((r) => setTimeout(r, rateLimitRecoveryDelayMs()));
  await timedCall(result, "Verify recovery", async () => {
    return await pds.accounts.getSession(luna.accessJwt);
  });

  result.finish();
  return result;
}

if (import.meta.main) {
  run().then((res) => {
    console.log(res.summary());
    Deno.exit(res.ok ? 0 : 1);
  });
}
